llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,492 @@
+      auto *CAM = cast<ConstantAsMetadata>(N.getOperand(I));
+      Constant *C = CAM->getValue();
+      Aspects.insert(cast<ConstantInt>(C)->getSExtValue());
+    }
+  }
+
+  return Result;
+}
+
+using TypesEdgesTy =
+    std::unordered_map<const Type *, std::vector<const Type *>>;
+
+/// Propagates aspects from type @Start to all types which
+/// are reachable by edges @Edges by BFS algorithm.
+/// Result is recorded in @Aspects.
+void propagateAspectsThroughTypes(const TypesEdgesTy &Edges, const Type *Start,
+                                  TypeToAspectsMapTy &Aspects) {
+  const AspectsSetTy &AspectsToPropagate = Aspects[Start];
+  SmallSetVector<const Type *, 16> TypesToPropagate;
+  TypesToPropagate.insert(Start);
+  for (size_t I = 0; I < TypesToPropagate.size(); ++I) {
+    const Type *T = TypesToPropagate[I];
+    Aspects[T].insert(AspectsToPropagate.begin(), AspectsToPropagate.end());
+    auto It = Edges.find(T);
+    if (It != Edges.end())
+      TypesToPropagate.insert(It->second.begin(), It->second.end());
+  }
+}
+
+/// Propagates given aspects to all types in module @M. Function accepts
+/// aspects in @TypesWithAspects reference and writes a result in this
+/// reference.
+/// Type T in the result contains an aspect A if Type T is a composite
+/// type (array, struct, vector) which contains elements/fields of
+/// another type TT, which in turn uses the aspect A.
+/// @TypesWithAspects argument consist of known types with aspects
+/// from metadata information.
+///
+/// The algorithm is the following:
+/// 1) Make a list of all structure types from module @M. The list also
+///    contains DoubleTy since it is optional as well.
+/// 2) Make from list a type graph which consists of nodes corresponding to
+///    types and directed edges between nodes. An edge from type A to type B
+///    corresponds to the fact that A is contained within B.
+///    Examples: B is a pointer to A, B is a struct containing field of type A.
+/// 3) For every known type with aspects propagate it's aspects over graph.
+///    Every propagation is a separate run of BFS algorithm.
+///
+/// Time complexity: O((V + E) * T) where T is the number of input types
+/// containing aspects.
+void propagateAspectsToOtherTypesInModule(
+    const Module &M, TypeToAspectsMapTy &TypesWithAspects) {
+  std::unordered_set<const Type *> TypesToProcess;
+  Type *DoubleTy = Type::getDoubleTy(M.getContext());
+
+  // 6 is taken from sycl/include/CL/sycl/aspects.hpp
+  static constexpr int AspectFP64 = 6;
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes())
+    TypesToProcess.insert(T);
+
+  TypesEdgesTy Edges;
+  for (const Type *T : TypesToProcess) {
+    for (const Type *TT : T->subtypes()) {
+      // If TT = %A*** then we want to get TT = %A
+      // The same with arrays and vectors
+      while (TT->isPointerTy() || TT->isArrayTy() || TT->isVectorTy()) {
```

Comment:
Typed pointers should be replaced by opaque pointers soon.
Could you please check that this code works for opaque pointers as well?

Reply:
I don't see that any of those APIs is going to be removed and therefore this code should continue to compile even with opaque pointers and it will continue to work correctly for arrays and vectors.

Pointer "branch" of this code won't work as it used to, because I assume there would be no contained type in it anymore.

@maksimsab, I think we should do the following:
```suggestion
      while ((TT->isPointerTy() && !TT->isOpaquePointerTy()) || TT->isArrayTy() || TT->isVectorTy()) {
```

This will skip opaque pointers and they will be handled elsewhere when they are accessed and this code will work for non-opaque pointers as intended by looking through them.

Reply:
In fact, I realized that opaque pointers might break the current solution because the first step of algorithm (which traverses over type graph) becomes insufficient for overall correctness.